dwh/indicator_lib.py

```diff
@@ -83,14 +83,54 @@ def calc_TX_PVLS(patient_agg_obs: pandas.DataFrame, VL_code: str,
   # Note the above copy is used to avoid setting a new column on a slice next:
-  temp_df['sup_VL'] = (temp_df['max_value'] < failure_threshold)
+  # separate value from date
+  temp_df[['vl_max_date', 'latest_vl_value']] = temp_df.max_date_value.str.split(query_lib.DATE_VALUE_SEPARATOR,
+                                                                                       expand=True)
+  temp_df['latest_vl_value'] = temp_df.loc[:, 'latest_vl_value'].astype(float)
+
+  temp_df['sup_VL'] = (temp_df.loc[:, 'latest_vl_value'] < failure_threshold)
   temp_df['buckets'] = temp_df.apply(
-      lambda x: _agg_buckets(x.birthDate, x.gender, datetime.today()), axis=1)
+      lambda x: _agg_buckets(x.birthDate, x.gender, end_date), axis=1)
   temp_df_exp = temp_df.explode('buckets')
   temp_df_exp = temp_df_exp.groupby(['sup_VL', 'buckets'], as_index=False)\
-      .count()[['sup_VL', 'buckets', 'patientId']]\
-      .rename(columns={'patientId': 'count'})
+      .count().loc[:, ['sup_VL', 'buckets', 'patientId']]\
+      .rename(columns={'patientId': 'sup_VL_count'})
   # calculate ratio
   num_patients = len(temp_df.index)
-  temp_df_exp['ratio'] = temp_df_exp['count']/num_patients
+  temp_df_exp['sup_VL_ratio'] = temp_df_exp.loc[:, 'sup_VL_count']/num_patients
   return temp_df_exp
+
+
+def calc_TX_NEW(patient_agg_obs:  pandas.DataFrame, ARV_plan: str, start_drug: List,
+    end_date_str: str = None) -> pandas.DataFrame:
+  """Calculates TX_NEW indicator with its corresponding disaggregations.
+
+  TX_NEW indicator counts the number of adults and children newly enrolled
+  on antiretroviral therapy (ART) within the provided
+
+  Args:
+    patient_agg_obs: A DataFrame generated by `patient_query.find_patient_aggregates()`.
+    ARV_plan: The concept question code for ANTIRETROVIRAL PLAN
+    start_drug: The concept answer code for START DRUG
+    end_date: The string representation of the last date as 'YYYY-MM-DD'.
+  Returns:
+  """
+  end_date = datetime.today()
+  if end_date_str:
+    end_date = datetime.strptime(end_date_str, '%Y-%m-%d')
+  temp_df = patient_agg_obs[(patient_agg_obs['code'] == ARV_plan)].copy()
+  # Note the above copy is used to avoid setting a new column on a slice next:
+  # separate value from date
+  temp_df[['ARV_max_date', 'latest_ARV_plan']] = temp_df.max_date_value_code.str.split(query_lib.DATE_VALUE_SEPARATOR,
+                                                                                  expand=True)
+  temp_df['TX_NEW'] = (temp_df.loc[:, 'latest_ARV_plan'].isin(start_drug))
+  temp_df['buckets'] = temp_df.apply(
+    lambda x: _agg_buckets(x.birthDate, x.gender, end_date), axis=1)
+  temp_df_exp = temp_df.explode('buckets')
+  temp_df_exp = temp_df_exp.groupby(['TX_NEW', 'buckets'], as_index=False) \
+    .count().loc[:, ['TX_NEW', 'buckets', 'patientId']] \
+    .rename(columns={'patientId': 'TX_NEW_count'})
+  # calculate ratio
+  num_patients = len(temp_df.index)
```

Comment:
Another question: Have you considered the case where the same `ARV_plan` code may appear in multiple observations of the same patient with different `valueCoding` (i.e., different `value.codeableConcept.coding` in FHIR)? Note our current aggregation on the Spark side includes `valueCoding` in `groupBy` ([here](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/2c9cc174ef47ea718de3a3c7c935e715cd3936cd/dwh/query_lib.py#L304)); should I change that?

Reply:
Good question. In general, if we have multiple observations with the same question mapped to multiple different answers (value.codeableConcept.coding), we might want to use the latest.

 We handled for this situation by introducing querying by max date, No?

Reply:
The way we handle this works for `value.quantity` but not `value.codeableConcept` and that's because of the `groupBy` issue I mentioned above. Take a look at this example (`valueCode` is always the same as the value suffix in `min_date_value_code`):

 
![image](https://user-images.githubusercontent.com/11869543/125104772-bf648f00-e0ab-11eb-825a-1789f4a5c634.png)

I can drop `valueCode` from `groupBy` but then we won't be able to differentiate the following two scenarios outside query library (is that okay?):

- In the query period, patient P has 3 observations with code C, with coded values V1, V2, V3, respectively.
- Patient Q has 2 observations with code C with coded values V1 and V3

In other words, outside the query library we only see the first and last observation values, even for coded values, hence we cannot tell if patient P has the V2 observation in this period or not. I think this should be fine but wanted to confirm with you since you are implementing the indicators.

Reply:
Thanks for making sure. This should be fine until we hit a need to tell if patient P has `V2 observation `